src/py21cmfast/src/PerturbField.c

```diff
@@ -102,103 +120,123 @@ int ComputePerturbField(float redshift, struct UserParams *user_params, struct C
+            for (i=0; i<user_params->DIM;i++){
+                for (j=0; j<user_params->DIM;j++){
+                    for (k=0; k<user_params->DIM;k++){
+                
+                        // map indeces to locations in units of box size
+                        xf = (i+0.5)/((user_params->DIM)+0.0);
+                        yf = (j+0.5)/((user_params->DIM)+0.0);
+                        zf = (k+0.5)/((user_params->DIM)+0.0);
+                    
+                        // update locations
+                        HII_i = (unsigned long long)(i/f_pixel_factor);
+                        HII_j = (unsigned long long)(j/f_pixel_factor);
+                        HII_k = (unsigned long long)(k/f_pixel_factor);
+                        xf += (boxes->lowres_vx)[HII_R_INDEX(HII_i, HII_j, HII_k)];
+                        yf += (boxes->lowres_vy)[HII_R_INDEX(HII_i, HII_j, HII_k)];
+                        zf += (boxes->lowres_vz)[HII_R_INDEX(HII_i, HII_j, HII_k)];
+                    
+                        // 2LPT PART
+                        // add second order corrections
+                        if(global_params.SECOND_ORDER_LPT_CORRECTIONS){
+                            xf -= (boxes->lowres_vx_2LPT)[HII_R_INDEX(HII_i,HII_j,HII_k)];
+                            yf -= (boxes->lowres_vy_2LPT)[HII_R_INDEX(HII_i,HII_j,HII_k)];
+                            zf -= (boxes->lowres_vz_2LPT)[HII_R_INDEX(HII_i,HII_j,HII_k)];
+                        }
+                    
+                        xf *= (float)(user_params->HII_DIM);
+                        yf *= (float)(user_params->HII_DIM);
+                        zf *= (float)(user_params->HII_DIM);
+                        while (xf >= (float)(user_params->HII_DIM)){ xf -= (user_params->HII_DIM);}
+                        while (xf < 0){ xf += (user_params->HII_DIM);}
+                        while (yf >= (float)(user_params->HII_DIM)){ yf -= (user_params->HII_DIM);}
+                        while (yf < 0){ yf += (user_params->HII_DIM);}
+                        while (zf >= (float)(user_params->HII_DIM)){ zf -= (user_params->HII_DIM);}
+                        while (zf < 0){ zf += (user_params->HII_DIM);}
+                        xi = xf;
+                        yi = yf;
+                        zi = zf;
+                        if (xi >= (user_params->HII_DIM)){ xi -= (user_params->HII_DIM);}
+                        if (xi < 0) {xi += (user_params->HII_DIM);}
+                        if (yi >= (user_params->HII_DIM)){ yi -= (user_params->HII_DIM);}
+                        if (yi < 0) {yi += (user_params->HII_DIM);}
+                        if (zi >= (user_params->HII_DIM)){ zi -= (user_params->HII_DIM);}
+                        if (zi < 0) {zi += (user_params->HII_DIM);}
+                    
+                        *( (float *)LOWRES_density_perturb + HII_R_FFT_INDEX(xi, yi, zi) ) +=
+                            (1 + init_growth_factor*(boxes->hires_density)[R_INDEX(i,j,k)]);
```

Comment:
One more thing (sorry @steven-murray !)...

I believe this line is technically a race condition if `LOWRES_density_perturb` is written to in the same location in memory by multiple threads at the same time (i.e. multiple hires cells being displaced to the same lowres location at the same time). I don't remember exactly what the consequences were (segfaults, probably?), but I do remember it caused me many headaches in the past in my old parallel implementation of 21cmFASTv1.

My workaround was to place the `omp  for` pragma at the `j` loop instead, store the computed `HII_R_FFT_INDEX(xi, yi, zi)` (at fixed `i`) in a separate 2D array since that is the most computationally intensive part, and then (serially) add the displaced mass to the 3D field. I'm sure there is a better way to do it with OpenMP structures, though.

Reply:
Perhaps this could be fixed by an `atomic` statement before the line?

Reply:
@steven-murray Yes, I think an atomic statement would suffice, but I can think about it a little further. 

Nice catch @freddavies! In this instance, no segfaults would arise, I think just the cells that incur the behaviour would have less density than they should.

Reply:
@steven-murray an atomic statement does fix this issue. Theoretically though this can cause rounding errors, owing to the fact that it is a float operation. However, you need to be down-sampling quite significantly to incur this. Down-sampling by > 20 did it, though its only at the 3rd-4th decimal place. As soon as it goes to the nominal factor of 3-4 its completely gone. 

The only true solution would to have this down-sampled grid allocated as a double and then convert to float. It would increase the memory, but not substantially. However, I am not sure its worth doing.

Actually, I might just go ahead and do it.